File: cplusplus/031_Next_Permutation.cpp
class Solution {
public:
  void nextPermutation(vector<int>& nums) {
    if (nums.size()<=1) return;
    int i = nums.size()-2;
    while (i>=0) {
      if (nums[i]<nums[i+1]) {
        break;
      } else {
        --i;
      }
    }
    std::reverse(nums.begin()+i+1,nums.end());
    if (i==-1) return;
    auto iter = std::upper_bound(nums.begin()+i+1,nums.end(),nums[i]);
    std::swap(nums[i],*iter);
  }
};

File: java/031_Next_Permutation.java
public class Solution {
    void reverseArray(int[] nums, int from) {
        int begin = from, end = nums.length - 1;
        while (begin < end) {
            int temp = nums[end];
            nums[end] = nums[begin];
            nums[begin] = temp;
            
            ++ begin;
            -- end;
        }
    }
    
    void swap(int[] nums, int left, int right) {
        int temp = nums[right];
        nums[right] = nums[left];
        nums[left] = temp;
    }
    
    public void nextPermutation(int[] nums) {
        if (nums == null || nums.length == 0) return;
        int index = nums.length - 1;
        while (index >= 1 && nums[index-1] >= nums[index]) {
            -- index;
        }
        -- index;
        reverseArray(nums, index + 1);
        if (index == -1) return;
        
        final int value = nums[index];
        for (int i = index + 1; i < nums.length; ++i) {
            if (nums[i] > value) {
                swap(nums, index, i);
                break;
            }
        }
    }
}
